rasa/core/domain.py

```diff
@@ -729,13 +742,30 @@ def as_dict(self) -> Dict[Text, Any]:
     def persist(self, filename: Union[Text, Path]) -> None:
         """Write domain to a file."""
 
-        domain_data = self.as_dict()
+        domain_data = self.domain_with_intents_transformed_for_file()
         utils.dump_obj_as_yaml_to_file(filename, domain_data)
 
-    def cleaned_domain(self) -> Dict[Text, Any]:
-        """Fetch cleaned domain, replacing redundant keys with default values."""
-
+    def domain_with_intents_transformed_for_file(self) -> Dict[Text, Any]:
+        """Replace the internal `used_entities` property by `use_entities` or `ignore_entities`."""
         domain_data = self.as_dict()
+        for idx, intent_info in enumerate(domain_data["intents"]):
+            for name, intent in intent_info.items():
+                use_entities = set(intent.pop("used_entities"))
+                ignore_entities = set(domain_data["entities"]) - use_entities
+                if len(use_entities) == len(domain_data["entities"]):
+                    intent["use_entities"] = True
+                elif len(use_entities) <= len(domain_data["entities"]) / 2:
+                    intent["use_entities"] = list(use_entities)
+                else:
+                    intent["ignore_entities"] = list(ignore_entities)
+        return domain_data
+
+    def cleaned_domain_for_file(self) -> Dict[Text, Any]:
```

Comment:
do we have to rename this method? there's downstream code in rasa-x relying on this, so if we do change it we need to make sure it's changed there as well. this might lead to unnecessary incompatibilities

Reply:
Thanks for pointing this out! I thought the new name is more descriptive, because "cleaned_domain" kind of implies that it is still a "domain" just cleaned up but here it has different properties after cleaning than an internally used domain. But the change isn't necessary, I'll change it back.